torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+    std::memcpy(data, key, len);
+  }
+  CacheKeyBuffer(CacheKeyBuffer&& other)
+      : data(std::exchange(other.data, nullptr)) {}
+  ~CacheKeyBuffer() {
+    if (data != nullptr) {
+      delete[] data;
+    }
+  }
+  CacheKeyBuffer(const CacheKeyBuffer& other) = delete;
+  CacheKeyBuffer& operator=(const CacheKeyBuffer&) = delete;
+
+  uint8_t* data;
+};
+
+struct CacheKey {
+  // Key to find the next node in the shadow graph.  We use C++ RTTI for the
+  // type of the node (ntype), then a key generated with a visitor pattern.
+  CacheKey(const std::type_index& ntype, const uint8_t* key, uint16_t len)
+      : node_type(ntype), key_size(len), key(key) {}
+
+  bool operator<(const CacheKey& other) const {
+    if (node_type != other.node_type) {
+      return node_type < other.node_type;
+    }
+    if (key_size != other.key_size) {
+      return key_size < other.key_size;
+    }
+    return std::memcmp(key, other.key, key_size) < 0;
+  }
+
+  bool operator==(const CacheKey& other) const {
+    return node_type == other.node_type && key_size == other.key_size &&
+        std::memcmp(key, other.key, key_size) == 0;
+  }
+
+  size_t hash() const {
+    // don't bother hashing the key data, common case 1 cache entry per node
+    return std::hash<std::type_index>()(node_type) ^ key_size;
+  }
+
+  std::type_index node_type;
+  uint16_t key_size;
+  const uint8_t* key;
+};
+
+struct OutputRef {
+  OutputRef() : node_id(-1), index(-1) {}
+  OutputRef(int node_, int index_) : node_id(node_), index(index_) {}
+
+  bool is_set() const {
+    return this->node_id >= 0;
+  }
+
+  int node_id;
+  int index;
+};
+
+struct NodeCall {
+  NodeCall(const std::shared_ptr<Node>& node_)
+      : node(node_), input_refs(node_->num_inputs()) {}
+
+  NodeCall(const std::shared_ptr<Node>& node_, int num_inputs_)
+      : node(node_), input_refs(num_inputs_) {}
+
+  OutputRef& operator[](size_t pos) {
+    return input_refs[pos];
+  }
+
+  size_t size() const {
+    return input_refs.size();
+  }
+
+  void mark_output(int input_nr, int output_idx) {
+    graph_output.emplace_back(std::make_pair(input_nr, output_idx));
+  }
+
+  std::shared_ptr<Node> node;
+  // at::SmallVector??
+  std::vector<OutputRef> input_refs;
+
+  // borrowed references
+  std::vector<std::pair<int, int>> tensor_pre_hooks;
+  std::vector<int> pre_hooks;
+  std::vector<int> post_hooks;
+  std::vector<std::pair<int, int>> graph_output;
+};
+
+struct AutogradCompilerCall {
+  AutogradCompilerCall(bool accumulate_grad_)
+      : accumulate_grad(accumulate_grad_),
+        default_dyn_type(SizeInput::STATIC) {}
+
+  void add_tensor_input(const at::Tensor& tensor) {
+    inputs.emplace_back(tensor);
+  }
+
+  void add_set_grad_target(const at::Tensor& tensor) {
+    set_grad_targets.emplace_back(tensor);
+  }
+
+  void add_size_input(const c10::SymInt& s) {
+    all_size_inputs.emplace_back(SizeInput(default_dyn_type, s.expect_int()));
```

Comment:
The SymInt is guaranteed to be concrete, because you only ever call this when doing real backwards (not when tracing?)

Reply:
I feel you may want to relax this constraint when we add support for capturing through backward(), since you'll want to run compiled autograd in the fake universe to generate an FX graph that dynamo can then continue to trace through. A comment is fine.

Reply:
Correct, compiled_autograd does not currently support fake tensors.  I agree we could extend it to do so.